File: PixelDungeon/src/main/java/com/nyrds/retrodungeon/mobs/common/Deathling.java
package com.nyrds.retrodungeon.mobs.common;

import com.nyrds.retrodungeon.mobs.necropolis.UndeadMob;
import com.watabou.pixeldungeon.Dungeon;
import com.watabou.pixeldungeon.actors.Char;
import com.watabou.utils.Random;

/**
 * Created by DeadDie on 20.08.2016
 */
public class Deathling extends UndeadMob {

    private static final int HEALTH = 4;

    public Deathling(){
        hp(ht(HEALTH + getModifier()));

        baseSpeed = 1.1f;
        defenseSkill = 1 + getModifier();
        flying = true;

        exp = 0;
        maxLvl = 32;
    }

    private int getModifier(){
        if (Dungeon.hero != null){
            return Dungeon.hero.lvl();
        }
        return 1;
    }

    @Override
    public int damageRoll() {
        return Random.NormalIntRange(1, 4 + getModifier());
    }

    @Override
    public int attackSkill( Char target ) {
        return 4 + getModifier();
    }

    @Override
    public int dr() {
        return 1 + getModifier();
    }
}

File: scripts/mobs/Deathling.lua
--[[
  Deathling - necromancy summon (batch 17c-2b,
  was com.nyrds.pixeldungeon.mobs.common.Deathling.java).
  A fallen hero's soul serving the summoner; its stats scale off the
  owner hero (lvl + skillLevel^2) and refresh every act. The first act
  after summoning fills the new hp total (firstAct, script data - rides
  the serpent save round-trip like Goo's pumpedUp). Summoned by the
  SummonDeathling spell, which makePet()s it.
  Accepted delta: the java ARTIFACT/LEFT_ARTIFACT equipment-slot override
  is dropped (pets are never equipped by the game).
]]
local RPD = require "scripts/lib/commonClasses"
local mob = require "scripts/lib/mob"

local HEALTH = 4

local function modifier(self)
    local owner = self:getOwner()
    if owner == nil or not owner:valid() then
        return 0
    end
    return owner:lvl() + owner:skillLevel() * owner:skillLevel()
end

return mob.init{
    act = function(self)
        local m = modifier(self)
        self:ht(HEALTH + m)

        local data = mob.restoreData(self)
        if not data.firstActDone then
            data.firstActDone = true
            self:heal(self:ht(), self)
        end
    end,

    defenseSkill = function(self, enemy)
        return 1 + modifier(self)
    end,

    attackSkill = function(self, target)
        return 4 + modifier(self)
    end,

    damageRoll = function(self)
        return math.random(1, 4 + modifier(self))
    end,

    dr = function(self)
        return modifier(self)
    end,

    spawn = function(self, level)
        self:setSkillLevel(3)
    end,
}
